src/Components/test/testassets/BasicTestApp/RouterTest/WithQueryParameters.razor

```diff
@@ -0,0 +1,32 @@
+﻿@page "/WithQueryParameters/{firstName}/{OptionalLastName?}"
+<strong id="test-info">Hello @FirstName @OptionalLastName.</strong>
+<p>IntValue: <strong id="value-QueryInt">@IntValue</strong></p>
+<p>NullableDateTimeValue: <strong id="value-NullableDateTimeValue">@NullableDateTimeValue?.ToString("hh:mm:ss on yyyy-MM-dd")</strong></p>
+<p>StringValue: <strong id="value-StringValue">@StringValue</strong></p>
+<p>LongValues: <strong id="value-LongValues">@LongValues.Length values (@string.Join(", ", LongValues.Select(x => x.ToString()).ToArray()))</strong></p>
+
+<p>Instance ID: <strong id="instance-id">@instanceId</strong></p>
+
+<p>
+    Links:
+    <a href="WithQueryParameters/@FirstName?intvalue=123">With IntValue</a> |
+    <a href="WithQueryParameters/@FirstName?intvalue=123&NullableDateTimeValue=@(new DateTime(2000, 1, 2, 3, 4, 5, 6).ToString("u"))">With NullableDateTimeValue</a> |
+    <a href="WithQueryParameters/@FirstName?l=50&l=100&l=-20&intvalue=123">With IntValue and LongValues</a> |
+</p>
+
+@code
+{
+    private string instanceId = Guid.NewGuid().ToString();
+
+    [Parameter] public string FirstName { get; set; }
+
+    [Parameter] public string OptionalLastName { get ; set; }
+
+    [Parameter, SupplyParameterFromQuery] public int IntValue { get ; set; }
+
+    [Parameter, SupplyParameterFromQuery] public DateTime? NullableDateTimeValue { get ; set; }
+
+    [Parameter, SupplyParameterFromQuery] public string StringValue { get ; set; }
+
+    [Parameter, SupplyParameterFromQuery(Name = "l")] public long[] LongValues { get ; set; }
+}
```

Comment:
Spicy question:

Is there a case where we have a parameter that can come from the route and that also has SupplyParameterFromQuery? 

While I get that the behavior might seem like an edge case (specially in Blazor) I think we should define a behavior for it that it is potentially consistent with ASP.NET Core/MVC model binding to avoid surprises.

Reply:
@javiercn Good question. I missed following up on this yesterday so will do now.

> Is there a case where we have a parameter that can come from the route and that also has SupplyParameterFromQuery?

Yes, that's possible. The behavior as of this PR is that we'll supply both values for the same parameter. That is, there will be two entries in the `ParameterView` matching this parameter name, and it's up to the `IComponent`'s `SetParametersAsync` to do whatever it wants with them both. `ParameterView` is not a dictionary and has always been able to contain multiple values for any given parameter name.

The default `SetParametersAsync` logic in `ComponentBase` will set them in the order they appear in the `ParameterView`, which happens to be "route parameters first, then query parameters", so the query parameter value will overwrite the route parameter value. People who want to customize this could implement `SetParametersAsync` manually, though not many will do.

In that sense it is already in a definite order. However TBH I don't see many cases where it's helpful to lose information by overwriting route values with query values. **If we're erring on the risk-averse side here, it would be most risk-averse for us to outlaw this whole scenario by throwing if a component declares query params that clash with route params.** Then we could always allow it in the future, but don't necessarily have to.

If we do want to throw in that case, I think we could do it by making `QueryParameterValueSupplier` directly know about the routing system and ask it for the list of routes matching the target component, and then walk them to see if there are any matching parameter names. It's not the most convenient thing to do but we can do it if we think it's the behavior we want.

@javiercn What are your opinions here?

Reply:
I don't have a strong opinion here:
* Matching what ASP.NET does is nice because because it builds on existing knowledge, but I don't think/feel it is required, specially if we have to couple this to the router.
  * Could it be done checking the current set of parameters in the ParameterView to make a decission and avoid searching for that parameter if its present on the existing provided values?
* I'm fine if we decide to leave it as is, it would be nice if we have a test to cover it. I agree with you that in most cases it doesn't add value in the context of Blazor.
* Optionally, if we find a cheap way to throw that doesn't involve coupling with routing (or any other parent that adds parameters for that matter), I would be ok with that too.

Those are my thoughts, however I fully trust you to make the right call here, I just thought it was worth raising it up to make sure we are ok with whatever behavior we choose here.

Reply:
> Could it be done checking the current set of parameters in the ParameterView

Yes, but then it would have to be done during every navigation, whereas if we based it on the declared route parameters for the target component, we could do it just once up front when we already cache the `QueryParameterValueSupplier`.

I don't personally have any philosophical objection to coupling querystring handling to the notion of `RouteAttribute`. If someone isn't using our router, then they either don't use `[Route]` in which case we can't anticipate their goals and it's up to them to implement whatever they want, or they do still use `[Route]` in which case it's very likely they want the same no-clash rule.

I'm also OK with just leaving this behavior alone because people won't have any reason to put themselves in this situation, and even if they do, it's not like it creates some new security issue (both route values and query values are equally untrusted).